packages/block-editor/src/components/block-list/index.js

```diff
@@ -49,22 +50,37 @@ const pendingBlockVisibilityUpdatesPerRegistry = new WeakMap();
 function Root( { className, ...settings } ) {
 	const [ element, setElement ] = useState();
 	const isLargeViewport = useViewportMatch( 'medium' );
-	const { isOutlineMode, isFocusMode, editorMode } = useSelect(
-		( select ) => {
-			const { getSettings, __unstableGetEditorMode } =
-				select( blockEditorStore );
-			const { outlineMode, focusMode } = getSettings();
-			return {
-				isOutlineMode: outlineMode,
-				isFocusMode: focusMode,
-				editorMode: __unstableGetEditorMode(),
-			};
-		},
-		[]
-	);
+	const {
+		isOutlineMode,
+		isFocusMode,
+		editorMode,
+		displayPrompt,
+		removalFunction,
+		blockToRemove,
+	} = useSelect( ( select ) => {
+		const {
+			getSettings,
+			__unstableGetEditorMode,
+			isRemovalPromptDisplayed,
+		} = select( blockEditorStore );
+		const { outlineMode, focusMode } = getSettings();
+		const {
+			displayPrompt: _displayPrompt,
+			removalFunction: _removalFunction,
+			blockName,
+		} = isRemovalPromptDisplayed();
+		return {
+			isOutlineMode: outlineMode,
+			isFocusMode: focusMode,
+			editorMode: __unstableGetEditorMode(),
+			displayPrompt: _displayPrompt,
+			removalFunction: _removalFunction,
+			blockToRemove: blockName,
```

Comment:
See my other comment about the need for a removal function. If we could distill the data needs here, I think we could reduce this to a nicer:

```suggestion
			shouldDisplayBlockRemovalPrompt: !! blockToBeRemoved(),
```

Where `select( 'core/block-editor' ).blockToBeRemoved` is the successor to `isRemovalPromptDisplayed` and only returns the block name if there is one, or `null/undefined`.

Reply:
Looking at this again, I wonder if this state should even be read/handled by `BlockList` at all. Shouldn't `BlockRemovalWarningModal` just handle everything internally? We have many such components:

```js
		<>
			<elementContext.Provider value={ element }>
				<IntersectionObserver.Provider value={ intersectionObserver }>
					<div { ...innerBlocksProps } />
				</IntersectionObserver.Provider>
			</elementContext.Provider>
			<BlockRemovalWarningModal />
		</>

```

where `BlockRemovalWarningModal` might become `BlockRemovalPromptListener` or anything that better conveys its conditional nature.

Reply:
Yeah good point, it'll be cleaner if the modal itself gets what it needs from state. As it stands, the modal needs to know both the name of the block to be removed, and the function to call if the user confirms the action.

Reply:
> Shouldn't BlockRemovalWarningModal just handle everything internally?

I've been playing with this a bit but can't quite get it to work, I think because if there's no state change in the block list when a block deletion is attempted, there's no reason to re-render the component so the modal never gets triggered when it's needed. Plus it renders a bunch of times when the block list first loads which is kind of annoying 😅

Reply:
I gave this a go in f0247fe65a38229afa5c28ccfc6f4858bdc6a539. This works, right?

Reply:
Yes! I was pretty much trying to do that, so not sure what went wrong 😕 probably had a typo somewhere. Thanks!